libs/cardano-data/src/Data/MapExtras.hs

```diff
@@ -126,11 +111,29 @@ disjointMapSetFold accum (Bin _ k v l1 l2) set !ans =
   where
     (s1, found, s2) = Set.splitMember k set
     addKV k1 v1 !ans1 = if not found then accum k1 v1 ans1 else ans1
+{-# INLINEABLE disjointMapSetFold #-}
 
 -- =================================
 
 intersectWhen1 :: Ord k => (k -> u -> v -> Bool) -> Map k u -> Map k v -> Map k u
 intersectWhen1 p x y = Map.mergeWithKey (\k u v -> if p k u v then Just u else Nothing) (const Map.empty) (const Map.empty) x y
+{-# INLINE intersectWhen1 #-}
 
 intersectWhen2 :: Ord k => (k -> u -> v -> Bool) -> Map k u -> Map k v -> Map k v
 intersectWhen2 p x y = Map.mergeWithKey (\k u v -> if p k u v then Just v else Nothing) (const Map.empty) (const Map.empty) x y
+{-# INLINE intersectWhen2 #-}
+
+-- | Partition the `Map` according to keys in the `Set`. This is equivalent to:
+--
+-- > extractKeysSet m s === (withoutKeys m s, restrictKeys m s)
+extractKeys :: Ord k => Map k a -> Set k -> (Map k a, Map k a)
+extractKeys sm = Set.foldl' f (sm, Map.empty)
+  where
+    f acc@(without, restrict) k =
+      case Map.lookup k without of
+        Nothing -> acc
+        Just v ->
+          let !without' = Map.delete k without
+              !restrict' = Map.insert k v restrict
+           in (without', restrict')
+{-# INLINEABLE extractKeys #-}
```

Comment:
This may have pretty bad complexity, since it visits every node in 'sm' .  Can't we use the property that the set and map are balanced using the same Ord instance, and split them both. If one of the splittings has a left or right side that is empty, we don't have to explore the (left ir right side) of the other?

Reply:
Yes we could! Thank you for pointing it out, @TimSheard 

I will go ahead an implement it, but I think I'll better do it in a separate PR, since I would prefer such change to be more visible. Also adding a couple of tests for this would be a good idea.

As far as this PR is concerned this function just moved from one module to another.

Reply:
OK, just don't forget.